File: week03/作业44-序列化和反序列化二叉树/Solution.java
public class Solution {
	// Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder res = mySeri(root, new StringBuilder());
        return res.toString();
    }

    StringBuilder mySeri(TreeNode root,StringBuilder sb){
        if (root == null) {
            sb.append("null,");
            return sb;
        } else if(root != null) {
            sb.append(root.val);
            sb.append(",");

            mySeri(root.left, sb);
            mySeri(root.right, sb);
        }
        return sb;
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] temp = data.split(","); // 将序列化的结果转为字符串数组
        List<String> list = new LinkedList<>(Arrays.asList(temp)); // 字符串数组转为集合类 便于操作
        return myDeSeri(list);
    }

    public TreeNode myDeSeri(List<String> list) {
        TreeNode root;
        if (list.get(0).equals("null")) {
            list.remove(0); // 删除第一个元素 则第二个元素成为新的首部 便于递归
            return null;
        } else {
            root=new TreeNode(Integer.valueOf(list.get(0)));
            list.remove(0);
            root.left = myDeSeri(list);
            root.right = myDeSeri(list);
        }
        return root;
    }
}

File: week03/作业44-序列化和反序列化二叉树/Solution.py
class Solution:
	def serialize(self, root):
		if root == None:
			return []
		stack = []
		res = []
		while stack:
			node = stack.pop(0)
			if node == None:
				res.append(null)
			else:
				res.append(node.val)
				stack.append(node.left)
				stack.append(node.right)
		return res
	def deserialize(self, data):
		if data == [] || data[0] == 'null':
			return None
		root = TreeNode(data.pop(0))
		nodes = [root]
		while len(data) >= 2:
			node = nodes.pop(0)
			if node == None:
				continue
			num = data.pop(0)
			if num =='null':
				node.left = None
			else:
				node.left = TreeNode(num)
			num = data.pop[0]
			if num == 'null':
				node.right = None
			else:
				node.right = TreeNode(num)
			nodes.append(node.left)
			nodes.append(node.right)
		return root


